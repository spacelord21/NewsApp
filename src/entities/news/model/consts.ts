export const DEFAULT_VALUE_OF_NEWS_PER_PAGE = 10;
