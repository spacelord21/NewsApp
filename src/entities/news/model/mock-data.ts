export const mockNews = [
  {
    id: 1,
    description: 'some description',
    image:
      'https://static.wikia.nocookie.net/kingofthehill/images/5/59/Bobbeah.jpg/revision/latest/scale-to-width-down/250?cb=20200724170428',
    title: "that's title",
  },
  {
    id: 2,
    description: 'some description',
    image:
      'https://static.wikia.nocookie.net/kingofthehill/images/5/59/Bobbeah.jpg/revision/latest/scale-to-width-down/250?cb=20200724170428',
    title: "that's title",
  },
  {
    id: 3,
    description: 'some description',
    image:
      'https://static.wikia.nocookie.net/kingofthehill/images/5/59/Bobbeah.jpg/revision/latest/scale-to-width-down/250?cb=20200724170428',
    title: "that's title",
  },
  {
    id: 4,
    description: 'some description',
    image:
      'https://static.wikia.nocookie.net/kingofthehill/images/5/59/Bobbeah.jpg/revision/latest/scale-to-width-down/250?cb=20200724170428',
    title: "that's title",
  },
  {
    id: 5,
    description: 'some description',
    image:
      'https://static.wikia.nocookie.net/kingofthehill/images/5/59/Bobbeah.jpg/revision/latest/scale-to-width-down/250?cb=20200724170428',
    title: "that's title",
  },
];
