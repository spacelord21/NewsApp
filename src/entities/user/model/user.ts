import {createSlice} from '@reduxjs/toolkit';

export const user = createSlice({
  name: 'user',
});
