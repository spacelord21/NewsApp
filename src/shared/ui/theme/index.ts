import { theme } from "./theme";

export { styled } from "./styled";
export { theme } from "./theme";
export { AppThemeProvider } from "./provider";

export type Theme = typeof theme;
