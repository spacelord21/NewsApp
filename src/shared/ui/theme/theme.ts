const THEME_GRID_STEP = 8;

export const theme = {
  name: "theme",
  palette: {
    background: {
      primary: "#fff",
      secondary: "#CAF2FD",
      tertiary: "	#333333",
    },
    text: {
      primary: "#000",
      secondary: "#fff",
      tertiary: "#000",
    },
    accent: {
      "color-primary-100": "#FEDFCE",
      "color-primary-200": "#FEB79E",
      "color-primary-300": "#FC866E",
      "color-primary-400": "#F95849",
      "color-primary-500": "#F50F0F",
      "color-primary-600": "#D20A1C",
      "color-primary-700": "#B00724",
      "color-primary-800": "#8E0428",
      "color-primary-900": "#75022A",
      "color-success-100": "#EAFCD4",
      "color-success-200": "#CFF9AA",
      "color-success-300": "#ABEE7D",
      "color-success-400": "#86DE5A",
      "color-success-500": "#54C92A",
      "color-success-600": "#3AAC1E",
      "color-success-700": "#249015",
      "color-success-800": "#13740D",
      "color-success-900": "#086009",
      "color-info-100": "#CAF2FD",
      "color-info-200": "#97E1FC",
      "color-info-300": "#62C6F6",
      "color-info-400": "#3BAAED",
      "color-info-500": "#0080E2",
      "color-info-600": "#0063C2",
      "color-info-700": "#0049A2",
      "color-info-800": "#003483",
      "color-info-900": "#00256C",
      "color-warning-100": "#FFF9CC",
      "color-warning-200": "#FFF099",
      "color-warning-300": "#FFE667",
      "color-warning-400": "#FFDC41",
      "color-warning-500": "#FFCC02",
      "color-warning-600": "#DBAA01",
      "color-warning-700": "#B78B01",
      "color-warning-800": "#936C00",
      "color-warning-900": "#7A5700",
      "color-danger-100": "#FFD8D7",
      "color-danger-200": "#FFB0B5",
      "color-danger-300": "#FF889C",
      "color-danger-400": "#FF6B92",
      "color-danger-500": "#FF3A82",
      "color-danger-600": "#DB2A7C",
      "color-danger-700": "#B71D74",
      "color-danger-800": "#931268",
      "color-danger-900": "#7A0B5F",
    },
  },
  typography: {
    title: {
      size: "24px",
      fontFamily: "ROBOTO_LIGHT",
      letterSpacing: "0px",
    },
    subtitle: {
      size: "16px",
      fontFamily: "ROBOTO_MEDIUM",
      letterSpacing: "0.15px",
    },
    body16: {
      size: "16px",
      fontFamily: "ROBOTO_REGULAR",
      letterSpacing: "0.5px",
    },
    caption: {
      size: "12px",
      fontFamily: "ROBOTO_LIGHT",
      letterSpacing: "0.4px",
    },
    newsTitle: {
      size: "26px",
      fontFamily: "ROBOTO_LIGHT",
      letterSpacing: "0.4px",
    },
  },
  spacing: (multiplier: number) => THEME_GRID_STEP * multiplier,
};
