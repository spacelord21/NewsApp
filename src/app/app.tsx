import { NavigationContainer } from "@react-navigation/native";
import { SafeAreaProvider, SafeAreaView } from "react-native-safe-area-context";
import { RootStack } from "./navigation/root-stack";
import { AppThemeProvider } from "@shared/ui";
import { Provider } from "react-redux";
import { persistor, store } from "./store";
import { PersistGate } from "redux-persist/integration/react";

export const App = () => {
  return (
    <SafeAreaProvider>
      <SafeAreaView style={{ flex: 1 }} edges={{ top: "maximum" }}>
        <AppThemeProvider>
          <Provider store={store}>
            <PersistGate persistor={persistor}>
              <NavigationContainer>
                <RootStack />
              </NavigationContainer>
            </PersistGate>
          </Provider>
        </AppThemeProvider>
      </SafeAreaView>
    </SafeAreaProvider>
  );
};
