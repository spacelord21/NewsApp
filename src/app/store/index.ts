import { newsReducer } from "@entities/news";
import { user } from "@entities/user";
import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { useDispatch } from "react-redux";
import AsyncStorage from "@react-native-async-storage/async-storage";
import {
  FLUSH,
  PAUSE,
  PERSIST,
  PURGE,
  PersistConfig,
  REGISTER,
  REHYDRATE,
  persistReducer,
  persistStore,
} from "redux-persist";
import thunk from "redux-thunk";

const persistConfig = {
  key: "root",
  storage: AsyncStorage,
  whitelist: ["user"],
};

export const rootReducer = combineReducers({
  newsReducer: newsReducer.reducer,
  user: user.reducer,
});

export const store = configureStore({
  reducer: persistReducer(persistConfig, rootReducer),
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }).concat(thunk),
});
export const persistor = persistStore(store);
export type AppDispatch = typeof store.dispatch;
export const useAppDispatch: () => AppDispatch = useDispatch;
