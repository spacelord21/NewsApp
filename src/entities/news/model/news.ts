import { createSelector, createSlice } from "@reduxjs/toolkit";
import { TNews } from "../types";
import { useSelector } from "react-redux";
import { fetchMoreNews, fetchNews } from "../api";
import { mappedData } from "../api/fetch-news/mapper";
import { RootState } from "@app/store/@types";

/* 
  т.к в предоставленном api нет возможности постраничной загрузки новостей, 
  я сделаю искусственный infinity scroll, изначально при авторизации загрузятся
  все новости, но с помощью redux'a я буду менять значение page и с таймаутом дополнять
  список отображаемых новостей 
*/

const DEFAULT_VALUE_OF_NEWS_PER_PAGE = 10;

export type TNewsSlice = {
  news: TNews[];
  loading: boolean;
  errorMessage: string;
  displayedNews: TNews[];
  amountOfPages: number;
};

export const initialState: TNewsSlice = {
  news: [],
  errorMessage: "",
  loading: false,
  displayedNews: [],
  amountOfPages: 0,
};

export const newsReducer = createSlice({
  name: "news",
  initialState: initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchNews.pending, (state, payload) => {
      return { ...state, loading: true };
    });
    builder.addCase(fetchNews.rejected, (state, { payload }) => {
      return { ...state, loading: false, errorMessage: payload! };
    });
    builder.addCase(fetchNews.fulfilled, (state, { payload }) => {
      return {
        ...state,
        amountOfPages: Math.ceil(
          payload.length / DEFAULT_VALUE_OF_NEWS_PER_PAGE
        ),
        loading: false,
        errorMessage: "",
        news: mappedData(payload),
        displayedNews: mappedData(
          payload.slice(0, DEFAULT_VALUE_OF_NEWS_PER_PAGE)
        ),
      };
    });
    builder.addCase(fetchMoreNews.pending, (state, _) => {
      return { ...state, loading: true };
    });
    builder.addCase(fetchMoreNews.fulfilled, (state, { payload }) => {
      const newsLength = state.news.length;
      const currentDisplayedNews = state.displayedNews;
      const displayedNewsLength = currentDisplayedNews.length;
      const nextSlice =
        payload * DEFAULT_VALUE_OF_NEWS_PER_PAGE > newsLength
          ? newsLength
          : payload * DEFAULT_VALUE_OF_NEWS_PER_PAGE;
      return {
        ...state,
        loading: false,
        displayedNews: [
          ...currentDisplayedNews,
          ...state.news.slice(displayedNewsLength, nextSlice),
        ],
      };
    });
  },
});

export const useNews = () =>
  useSelector(
    createSelector(
      (state: RootState) => state.newsReducer,
      (state) => {
        return state;
      }
    )
  );
