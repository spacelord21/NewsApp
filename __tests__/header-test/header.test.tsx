import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import { Header } from "../../src/widgets";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { NavigationContainer } from "@react-navigation/native";
import { logout, useUser } from "../../src/entities/user/model/user";
import { theme } from "../../src/shared/ui/theme/theme";
import { ThemeProvider } from "styled-components/native";
import { useAppDispatch } from "../../src/app/store";
import { useNavigation } from "@react-navigation/native";

jest.mock("@react-native-async-storage/async-storage", () =>
  require("@react-native-async-storage/async-storage/jest/async-storage-mock")
);

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: jest.fn(),
}));

jest.mock("../../src/app/store", () => ({
  useAppDispatch: jest.fn(),
}));

jest.mock("@react-navigation/native", () => ({
  ...jest.requireActual("@react-navigation/native"),
  useNavigation: jest.fn(),
}));

jest.mock("../../src/entities/user/model/user", () => ({
  useUser: jest.fn(),
  logout: jest.fn(),
}));

const middlewares = [];
const mockStore = configureStore(middlewares);

describe("Header component", () => {
  const mockUser = {
    avatar_url: "https://some.ru",
    username: "John Doe",
  };
  const store = mockStore({});

  it("информация о пользователе и автарка рендерятся корректно", () => {
    (useUser as jest.Mock).mockReturnValue(mockUser);
    const component = render(
      <ThemeProvider theme={theme}>
        <NavigationContainer>
          <Provider store={store}>
            <Header />
          </Provider>
        </NavigationContainer>
      </ThemeProvider>
    );

    expect(component.getByText("John Doe")).toBeTruthy();
    expect(component.getAllByTestId("avatar_comp")).toBeDefined();
    expect(component).toMatchSnapshot();
  });
  it("аватарки нет, если avatar_url пустой", () => {
    const mockUser = {
      avatar_url: "",
      username: "John Doe",
    };
    (useUser as jest.Mock).mockReturnValue(mockUser);
    const { queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <NavigationContainer>
          <Provider store={store}>
            <Header />
          </Provider>
        </NavigationContainer>
      </ThemeProvider>
    );
    expect(queryByTestId("avatar_comp")).toBeNull();
  });
  it("вызов navigation.navigate and dispatch on logout", () => {
    (useUser as jest.Mock).mockReturnValue(mockUser);
    const mockDispatch = jest.fn();
    const mockNavigation = jest.fn();
    (useAppDispatch as jest.Mock).mockReturnValue(mockDispatch);
    (useNavigation as jest.Mock).mockReturnValue({
      navigate: mockNavigation,
    });
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <NavigationContainer>
          <Provider store={store}>
            <Header />
          </Provider>
        </NavigationContainer>
      </ThemeProvider>
    );

    fireEvent.press(getByText("Выйти"));
    expect(mockNavigation).toHaveBeenCalledWith("authStack");
    expect(mockDispatch).toHaveBeenCalledWith(logout());
  });
});
