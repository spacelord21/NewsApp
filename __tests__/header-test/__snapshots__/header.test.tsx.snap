// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component информация о пользователе и автарка рендерятся корректно 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#333333",
        "flexDirection": "row",
        "height": 90,
        "justifyContent": "space-around",
        "width": "100%",
      },
    ]
  }
>
  <Text
    allowFontScaling={false}
    style={
      [
        {
          "color": "#fff",
          "fontFamily": "ROBOTO_MEDIUM",
          "fontSize": 16,
          "letterSpacing": 0.15,
        },
      ]
    }
    variant="subtitle"
  >
    John Doe
  </Text>
  <Image
    source={
      {
        "uri": "https://some.ru",
      }
    }
    style={
      [
        {
          "borderRadius": 50,
          "height": 70,
          "width": 70,
        },
      ]
    }
    testID="avatar_comp"
  />
  <Text
    allowFontScaling={false}
    onPress={[Function]}
    style={
      [
        {
          "color": "#fff",
          "fontFamily": "ROBOTO_REGULAR",
          "fontSize": 16,
          "letterSpacing": 0.5,
        },
      ]
    }
  >
    Выйти
  </Text>
</View>
`;
